plugins/unicodelib/unicodelib.cpp

```diff
@@ -278,6 +278,75 @@ inline unsigned char min3(unsigned char a, unsigned char b, unsigned char c)
 }
 
 #define DISTANCE_ON_ERROR 999
+class UPCList // User perceived character list
+{
+private:
+    UnicodeString ustring_;
+    uint32_t* c_offsets_;
+    uint32_t  length_;
+    uint32_t  capacity_;
+    bool invalid_;
+    void doCreateUPCList(BreakIterator& cbi) {
+        UErrorCode status = U_ZERO_ERROR;
+        if (!capacity_) {
+             capacity_ = ustring_.length();
+         }
+        c_offsets_ = new uint32_t[capacity_+1]; // the number of characters is always less or equal to the string length
+        unsigned index=0;
+        cbi.setText(ustring_);
+        c_offsets_[index] = cbi.first();
+        for (int32_t end = cbi.next(); end != BreakIterator::DONE && length_ < capacity_; end = cbi.next())
+        {
+            length_++;
+            c_offsets_[++index]=end;
+        }
+        if (U_FAILURE(status)) { length_ = 0; capacity_ = 0; invalid_ = true; }
+    }
+
+public:
+    UPCList(BreakIterator& cbi, const UnicodeString & source, uint32_t capacity=0)
+        : length_(0), capacity_(capacity),ustring_(source), invalid_(false)
+    {
+        doCreateUPCList(cbi);
+    }
+
+    ~UPCList()
+    {
+        delete[] c_offsets_;
+    }
+
+    uint32_t charOffset(uint32_t index) const
+    {
+        return (index < length_ )? c_offsets_[index]:0;
+    }
+
+    uint32_t charLength(uint32_t index) const
+    {
+        return (index < length_ )? c_offsets_[index+1]-c_offsets_[index]:0;
+    }
+
+    int compare(uint32_t index, const UPCList& srcText, uint32_t srcIndex) const
+    {
+        uint32_t lLen = charLength(index);
+        uint32_t rLen = srcText.charLength(srcIndex);
+        if ( lLen != rLen )
+            return (lLen < rLen) ? -1 : 1;
```

Comment:
Minor: slightly strange string ordering  'z' < 'aa'. I suspect it makes no difference.